File: r_csv/scrape_dsv.R
# scrape_dsv.R
# Purpose: Scrape a table from the Wikipedia page about Delimiter-separated values

# Load the rvest package (install it first if you haven’t)
# install.packages("rvest")

library(rvest)

# URL of the page
url <- "https://en.wikipedia.org/wiki/Delimiter-separated_values"

# Read the HTML
page <- read_html(url)

# Extract all tables on the page
tables <- html_nodes(page, "table")

# Find the first table with more than one column
chosen <- NULL
for (i in seq_along(tables)) {
  tbl <- tryCatch(html_table(tables[[i]], fill = TRUE), error = function(e) NULL)
  if (!is.null(tbl) && ncol(tbl) > 1) {
    chosen <- tbl
    break
  }
}

# Stop if no table found
if (is.null(chosen)) {
  stop("No suitable table found on the page")
}

# Create folder if not existing
dir.create("r_csv", showWarnings = FALSE)

# Save the table as CSV
write.csv(chosen, file = "r_csv/dsv_example.csv", row.names = FALSE)

# Optional: read back to confirm
verify <- read.csv("r_csv/dsv_example.csv", stringsAsFactors = FALSE)
print(head(verify))

File: python_csv/scrape_dsv.py
# scrape_dsv.py
# Purpose: Scrape a table from the Wikipedia page about Delimiter-separated values

import requests
from bs4 import BeautifulSoup
import pandas as pd
import os

# 1. Download the Wikipedia page
url = "https://en.wikipedia.org/wiki/Delimiter-separated_values"
headers = {"User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"}
resp = requests.get(url, headers=headers)

resp.raise_for_status()   # stops the script if download fails

# 2. Parse the HTML
soup = BeautifulSoup(resp.text, "html.parser")
all_tables = soup.find_all("table")

# 3. Find the first table that looks valid (more than one column)
chosen = None
for table in all_tables:
    try:
        df = pd.read_html(str(table))[0]
        if df.shape[1] > 1:
            chosen = df
            break
    except Exception:
        continue

if chosen is None:
    raise SystemExit("No suitable table found on the page")

# 4. Save it to a CSV file inside python_csv
os.makedirs("python_csv", exist_ok=True)
chosen.to_csv("python_csv/dsv_example.csv", index=False)

# 5. Verify by reading it back and printing first rows
verify = pd.read_csv("python_csv/dsv_example.csv")
print(verify.head())
